gds/src/main/scala/edu/gemini/aspen/gds/observations/ObservationFSM.scala

```diff
@@ -0,0 +1,115 @@
+import cats.effect.syntax.all._
+import cats.syntax.all._
+import edu.gemini.aspen.gds.syntax.all._
+import edu.gemini.aspen.giapi.data.{ DataLabel, ObservationEvent }
+import java.util.logging.Logger
+import scala.concurrent.duration._
+
+trait ObservationFSM[F[_]] {
+  def addObservationEvent(obsEvent: ObservationEvent): F[Unit]
+  def stopObservation: F[Unit]
+  def step: F[Unit]
+}
+
+object ObservationFSM {
+  private val logger = Logger.getLogger(this.getClass.getName)
+
+  // TODO: make these configurable
+  val numAttempts = 3
+  val sleepTime   = 5.seconds
+
+  def apply[F[_]: Async](
+    dataLabel: DataLabel,
+    obsStateQ: QueueSink[F, ObservationStateEvent]
+  ): F[ObservationFSM[F]] =
+    Ref.of[F, State](Running(Set.empty)).map { state =>
+      new ObservationFSM[F] {
+        val allEvents: Set[ObservationEvent] = ObservationEvent.values.toSet
+
+        def addObservationEvent(obsEvent: ObservationEvent): F[Unit] = state.modify {
+          case st @ Running(events) =>
+            if (events.contains(obsEvent))
+              st -> logger.warningF(
+                s"Duplicate event $obsEvent received for running observation $dataLabel"
+              )
+            else if (obsEvent == ObservationEvent.OBS_END_DSET_WRITE)
+              WaitingForEvents(events + obsEvent, numAttempts) -> (logger.infoF(
+                s"Received $obsEvent for observation $dataLabel"
+              ) >> qKeywordCollection(obsEvent) >> qStep)
+            else
+              Running(events + obsEvent)                       -> (logger.infoF(
+                s"Added $obsEvent to running observation $dataLabel"
+              ) >> qKeywordCollection(obsEvent))
+
+          case st @ WaitingForEvents(events, _) =>
+            if (events.contains(obsEvent))
+              st                                               -> logger.warningF(
+                s"Duplicate event $obsEvent received for observation $dataLabel, which was waiting for other events to complete."
+              )
+            else
+              WaitingForEvents(events + obsEvent, numAttempts) -> (logger.infoF(
+                s"Received $obsEvent for observation $dataLabel, which was waiting for events to complete."
+              ) >> qKeywordCollection(obsEvent) >> qStep)
+
+          case st @ _ =>
+            st -> logger.infoF(
+              s"Received $obsEvent for observation $dataLabel, which no longer needs events."
+            )
+        }.flatten
+
+        def stopObservation: F[Unit] = state.modify {
+          case Running(events) =>
+            WaitingForEvents(events, numAttempts) -> (logger.infoF(
+              s"Observation $dataLabel stopped by Seqexeq"
+            ) >> qStep)
+          case st @ _          => st -> Sync[F].unit
+        }.flatten
+
+        def step: F[Unit] = state.modify {
+          case st @ WaitingForEvents(_, _) => st -> waitForEvents
+          case st @ _                      => st -> Sync[F].unit
```

Comment:
`Applicative[F].unit` maybe? Just a small detail

Reply:
Good call. I changed from a context bound to an `implicit F: Async[F]` so I can just do `F.unit`. Would you prefer the `Applicative[F].unit` instead?